File: website/stories/forms/hypatia/hypatia.ts
import { hypatiaSettings } from "@/stories/blueprints/hypatia/hypatia";
import {
  DoubleSide,
  ExtrudeGeometry,
  IcosahedronGeometry,
  MathUtils,
  Mesh,
  MeshBasicMaterial,
  Path,
  Shape,
} from "three";

import { GUIType } from "@utils/gui/gui-types";
import { TWO_PI } from "@utils/three/math";
import WebGLApp, { GUIWebGLApp } from "../webgl-app";

export default class HypatiaSketch extends WebGLApp {
  up!: Mesh<ExtrudeGeometry, MeshBasicMaterial>;
  material = new MeshBasicMaterial({
    color: 0x333333,
    side: DoubleSide,
    wireframe: false,
  });
  wireframeMaterial = new MeshBasicMaterial({
    color: 0xffffff,
    wireframe: true,
    side: DoubleSide,
  });

  orbitMeshes: Mesh[] = [];

  outlineHorizontal!: Mesh<ExtrudeGeometry, MeshBasicMaterial>;
  outlineVerticalX!: Mesh<ExtrudeGeometry, MeshBasicMaterial>;
  outlineVerticalZ!: Mesh<ExtrudeGeometry, MeshBasicMaterial>;
  planets: Mesh<IcosahedronGeometry, MeshBasicMaterial>[] = [];

  radius = 500;
  ringSpeeds: number[] = [];

  animation = {
    speed: 1,
  };

  create() {
    this.cameras.main.position.z = 750;
    this.cameras.main.lookAt(0, 0, 0);

    // this.createHypatia();
    // this.createOutline();
    this.createOrbit();
  }

  createHypatia() {
    const mesh = new Mesh(
      new IcosahedronGeometry(
        this.radius * hypatiaSettings.form.hypatia.radius,
        3,
      ),
      new MeshBasicMaterial({ wireframe: false }),
    );
    mesh.name = "Hypatia";
    this.scene.add(mesh);
    return mesh;
  }

  createOutline() {
    const meshes = [];
    let total = 1;
    for (let i = 0; i < total; i++) {
      const theta = i * (TWO_PI / total);
      const r = this.radius / 2;

      const meshX = this.createRing(r, 1, 1, 0xffffff);
      const meshY = this.createRing(r, 1, 1, 0xffffff);
      const meshZ = this.createRing(r, 1, 1, 0xffffff);

      meshX.rotation.x = theta;
      meshY.rotation.y = theta;
      meshZ.rotation.y = theta;
      meshZ.rotation.x = Math.PI / 2;

      meshes.push(meshX, meshY, meshZ);
    }
  }

  createRing(
    radius: number,
    thickness: number,
    depth: number,
    color = 0xff0000,
  ) {
    // Create ring geometry with thin tube and current radius

    // Use RingGeometry to create the base ring shape
    // Create a shape for the ring by subtracting an inner circle from an outer circle
    const shape = new Shape();
    shape.absarc(0, 0, radius, 0, Math.PI * 2, false);

    // Create a hole (inner circle) to make it a ring
    const hole = new Path();
    hole.absarc(0, 0, radius - thickness, 0, Math.PI * 2, true);
    shape.holes.push(hole);

    const geometry = new ExtrudeGeometry(shape, {
      steps: 1,
      depth: depth,
      bevelEnabled: false,
      curveSegments: 64, // Increase the number of segments for smoother edges
    });

    const mesh = new Mesh(
      geometry,
      new MeshBasicMaterial({ color: color, wireframe: true }),
    );

    this.scene.add(mesh);

    return mesh;
  }

  createOrbit() {
    const total = 1;
    const thickness = 2;
    const r = this.radius / 2;
    const minRadius = r / 5;
    const maxRadius = r;

    const SM_AXIS_RATIO = 1 / Math.sqrt(3); // ~0.577

    for (let i = 0; i < total; i++) {
      const p = i / (total - 1) || 0;
      const radius = MathUtils.lerp(minRadius, maxRadius, p);

      const perspectiveFactorY = MathUtils.lerp(
        SM_AXIS_RATIO,
        SM_AXIS_RATIO * (1 + SM_AXIS_RATIO),
        p,
      );

      // Create ring geometry with thin tube and current radius

      // Use RingGeometry to create the base ring shape
      // Create a shape for the ring by subtracting an inner circle from an outer circle
      const shape = new Shape();
      shape.absellipse(
        0,
        0,
        radius,
        radius * perspectiveFactorY,
        0,
        Math.PI * 2,
        false,
      );

      // Create a hole (inner circle) to make it a ring
      const hole = new Path();
      // hole.absarc(0, 0, radius - thickness, 0, Math.PI * 2, true);
      hole.absellipse(
        0,
        0,
        radius - thickness,
        (radius - thickness) * perspectiveFactorY,
        0,
        Math.PI * 2,
        true,
      );
      shape.holes.push(hole);

      // Extrude the ring geometry to give it depth on the edges
      const extrudeSettings = {
        steps: 1,
        depth: thickness,
        bevelEnabled: false,
        curveSegments: 64, // Increase the number of segments for smoother edges
      };

      const geometry = new ExtrudeGeometry(shape, extrudeSettings);

      const ring = new Mesh(geometry, this.material);
      ring.add(new Mesh(geometry, this.wireframeMaterial));

      // ring.rotation.x = Math.PI / 2;
      let theta = MathUtils.lerp(0, 45, p); // Calculate current angle

      // ring.rotation.x = MathUtils.degToRad(theta); // Rotate ring to face camera
      // ring.rotation.x = MathUtils.degToRad(theta); // Rotate ring to face camera
      // ring.rotation.x = Math.PI / 2;
      ring.scale.z = perspectiveFactorY;

      this.ringSpeeds.push(MathUtils.lerp(2, 1, p)); // Randomize ring speeds

      this.scene.add(ring);

      this.orbitMeshes.push(ring); // Store each ring for later animation

      // Planet
      if (false) {
        const startAngle = Math.PI;
        theta = MathUtils.lerp(
          startAngle,
          startAngle + MathUtils.degToRad(hypatiaSettings.form.planets.spiral),
          p,
        );

        let mesh = new Mesh(
          new IcosahedronGeometry(
            radius * hypatiaSettings.form.planets.radius * 2,
            3,
          ),
          new MeshBasicMaterial({ wireframe: false, color: 0xffffff }),
        );
        mesh.position.set(
          Math.cos(theta) * radius,
          Math.sin(theta) * radius * perspectiveFactorY,
          0,
        );
        mesh.userData.theta = theta;
        mesh.userData.radius = radius;
        mesh.userData.perspectiveFactorY = perspectiveFactorY;
        mesh.name = `Planet-0-${i}`;
        ring.add(mesh);

        this.planets.push(mesh);

        theta =
          Math.PI +
          MathUtils.lerp(
            startAngle,
            startAngle +
              MathUtils.degToRad(hypatiaSettings.form.planets.spiral),
            p,
          );

        mesh = new Mesh(
          new IcosahedronGeometry(
            radius * hypatiaSettings.form.planets.radius * 2,
            3,
          ),
          new MeshBasicMaterial({ wireframe: false, color: 0xffffff }),
        );
        mesh.userData.theta = theta;
        mesh.userData.radius = radius;
        mesh.userData.perspectiveFactorY = perspectiveFactorY;
        mesh.position.set(
          Math.cos(theta) * radius,
          Math.sin(theta) * radius * perspectiveFactorY,
          0,
        );
        mesh.name = `Planet-1-${i}`;
        ring.add(mesh);

        this.planets.push(mesh);
      }
    }
  }

  onUpdate(delta: number) {
    return;
    this.planets.forEach((mesh: Mesh) => {
      mesh.userData.theta += delta;
      mesh.position.set(
        Math.cos(mesh.userData.theta) * mesh.userData.radius,
        Math.sin(mesh.userData.theta) *
          mesh.userData.radius *
          mesh.userData.perspectiveFactorY,
        0,
      );
    });

    this.orbitMeshes.forEach((ring, i) => {
      // ring.rotation.x += delta;
      ring.rotation.x += delta * this.animation.speed * this.ringSpeeds[i];
      // ring.rotation.z += delta;
    });
  }
}

/// #if DEBUG
export class GUIHypatiaSketch extends GUIWebGLApp {
  constructor(gui: GUIType, target: WebGLApp) {
    super(gui, target);
    this.gui = gui;
  }
}
/// #endif

File: website/stories/forms/hypatia/hypatia.py
import pymel.core as pm
import pymel.core.datatypes as dt
import math

RADIUS = 25.0
TORUS_SUBDIVISIONS = 200.0

# Outline
OUTLINE_RADIUS = 1
ORBIT_PROFILE_RADIUS = 0.5

# Hypatia
HYPATIA_RADIUS = 0.05

# Orbit
ORBIT_THICKNESS = 0.0025
ORBIT_HEIGHT = HYPATIA_RADIUS*2

# ----------------------------------------------------
# Utility Function: Linear Interpolation (lerp)
# ----------------------------------------------------
def lerp(a, b, t):
    """Linearly interpolate between a and b by t."""
    return a + (b - a) * t

 # Define the mapLinear equivalent function
def mapLinear(value, in_min, in_max, out_min, out_max):
    return out_min + (value - in_min) * (out_max - out_min) / (in_max - in_min)

def create_light_torus(major_radius=10, minor_radius=1, subdivisions_major=20, subdivisions_minor=10, name="regularTorus"):
    """
    Creates a regular torus mesh using the polyTorus command.

    Parameters:
      major_radius (float): The major (ring) radius of the torus.
      minor_radius (float): The minor (tube) radius of the torus.
      subdivisions_major (int): Number of subdivisions around the torus ring.
      subdivisions_minor (int): Number of subdivisions along the tube.
      name (str): The name for the torus mesh.

    Returns:
      PyNode: The created torus mesh.
    """
    torus_mesh = pm.polyTorus(r=major_radius, sr=minor_radius, sx=subdivisions_major, sy=subdivisions_minor, name=name)[0]
    pm.select(torus_mesh)
    return torus_mesh


def adjust_uvs_to_range(mesh, projection_scale_u=1, projection_scale_v=1):
    """
    Adjusts the UVs of a given mesh to ensure they fit within the 0-1 range using cylindrical mapping.
    
    Parameters:
    - mesh: The mesh to adjust the UVs for.
    - projection_scale_u: Scaling factor for U-axis in cylindrical projection (default is 1).
    - projection_scale_v: Scaling factor for V-axis in cylindrical projection (default is 1).
    """
    # Perform cylindrical projection
    pm.select(mesh)
    pm.polyCylindricalProjection(mesh, projectionScaleU=projection_scale_u, projectionScaleV=projection_scale_v, smartFit=True)
    
    # Now adjust the UV coordinates to ensure they fit in the 0-1 range.
    uvs = pm.polyListComponentConversion(mesh, toUV=True)
    uvs = pm.ls(uvs, flatten=True)

    # Calculate the range of UVs along U-axis
    min_u = float('inf')
    max_u = float('-inf')

    # Loop through each UV and calculate its U value
    for uv in uvs:
        u_value = pm.polyEditUV(uv, query=True, u=True)[0]
        min_u = min(min_u, u_value)
        max_u = max(max_u, u_value)

    # Apply mapLinear and clamp the values to the range [0, 1]
    for uv in uvs:
        u_value = pm.polyEditUV(uv, query=True, u=True)[0]
        mapped_u = mapLinear(u_value, min_u, max_u, 0.0, 1.0)
        clamped_u = max(0.0, min(1.0, mapped_u))  # Clamp the value to [0, 1]
        pm.select(uv)  # Select the individual UV
        pm.polyEditUV(pivotU=0.5, pivotV=0.5, uValue=clamped_u, relative=False)  # Apply the new UV value


# ----------------------------------------------------
# Function: Create a Closed Quadrant Profile
# ----------------------------------------------------
def create_quadrant_profile(circle_radius=1.0, cross_thickness=0.5, arc_sections=8):
    """
    Creates a fully closed quadrant curve profile.
    
    The profile:
      1. Starts at the bottom-left corner of the cross.
      2. Goes to the horizontal intersection.
      3. Follows an arc (using lerp for smooth spacing) from the horizontal to vertical intersection.
      4. Ends at the vertical intersection and closes back at the corner.
    """
    # Compute intersection points.
    x_intersect = math.sqrt(circle_radius**2 - cross_thickness**2)
    y_intersect = math.sqrt(circle_radius**2 - cross_thickness**2)
    
    pt_horizontal = (x_intersect, cross_thickness, 0)   # Intersection on horizontal bar.
    pt_vertical   = (cross_thickness, y_intersect, 0)    # Intersection on vertical bar.
    pt_corner     = (cross_thickness, cross_thickness, 0)  # Bottom-left corner.
    
    # Compute arc angles.
    angle_start = math.degrees(math.atan2(cross_thickness, x_intersect))
    angle_end   = math.degrees(math.atan2(y_intersect, cross_thickness))
    
    # Build list of points.
    points = []
    points.append(pt_corner)         # Start at corner.
    points.append(pt_horizontal)     # Go to horizontal intersection.
    
    # Generate arc points using lerp for smooth spacing.
    arc_points = []
    for i in range(arc_sections + 1):  # Include endpoint.
        angle = math.radians(lerp(angle_start, angle_end, i / arc_sections))
        x = math.cos(angle) * circle_radius
        y = math.sin(angle) * circle_radius
        arc_points.append((x, y, 0))
    points.extend(arc_points)
    
    points.append(pt_vertical)       # Move to vertical intersection.
    points.append(pt_corner)         # Close loop.
    
    # Create the curve.
    quadrant_curve = pm.curve(d=1, p=points, name="quadrantProfile")
    pm.makeIdentity(quadrant_curve, apply=True, translate=True, rotate=True, scale=True)
    pm.delete(quadrant_curve, constructionHistory=True)
    return quadrant_curve

# ----------------------------------------------------
# Function: Sweep a Profile Along a Circular Path
# ----------------------------------------------------
def sweep_profile(profile_curve, torus_radius=10, path_divisions=50):

    """
    Sweeps the provided profile curve along a circular path to create a torus-like mesh.
    
    The profile is moved to the torus edge and rotated so that it is perpendicular to the path.
    """
    # Create the circular sweep path.
    circle_path = pm.circle(radius=torus_radius, normal=(0, 1, 0), 
                            sections=path_divisions, name="torusPath")[0]
    
    # Align the profile: move it so its origin sits at the torus edge.
    pm.xform(profile_curve, translation=(torus_radius, 0, 0))
    # Rotate the profile so that it's perpendicular to the path.
    pm.xform(profile_curve, rotation=(0, 90, 0), worldSpace=True)
    
    # Perform the sweep (extrude in pipe mode).
    swept = pm.extrude(profile_curve, circle_path, et=2, upn=True, 
                       fixedPath=True, scale=1, useComponentPivot=True, 
                       name="sweptTorus")[0]
    
    # Cleanup the construction curves.
    pm.delete(profile_curve, circle_path)
    return swept


# ----------------------------------------------------
# MAIN PROCESS: Duplicate, Sweep, and Combine Quadrants
# ----------------------------------------------------
def create_motion_orbit_sculpture():
    circle_radius = OUTLINE_RADIUS*ORBIT_PROFILE_RADIUS
    # Create the original quadrant profile.
    original_profile = create_quadrant_profile(circle_radius=circle_radius, cross_thickness=circle_radius*0.25, arc_sections=8)
    
    swept_meshes = []
    # Define the rotation angles for the four quadrants.
    for angle in [0, 90, 180, 270]:
        # Duplicate the original quadrant profile.
        profile_dup = pm.duplicate(original_profile)[0]
        # Rotate it about the Z-axis by the specified angle.
        pm.xform(profile_dup, rotation=(0, 0, angle), relative=True, worldSpace=True)
        pm.makeIdentity(profile_dup, apply=True, translate=True, rotate=True, scale=True)
        # Sweep the rotated profile along the torus path.
        swept = sweep_profile(profile_dup, torus_radius=OUTLINE_RADIUS*RADIUS, path_divisions=TORUS_SUBDIVISIONS)
        # swept_meshes.append(swept)

        poly_mesh = pm.nurbsToPoly(swept, format=3, polygonType=1, name="polySweptTorus")[0]
        pm.polyNormal(poly_mesh, normalMode=0, userNormalMode=0)
        pm.delete(swept)
        swept_meshes.append(poly_mesh)
    
    
    # Combine all the swept meshes into one final torus.
    full_torus = pm.polyUnite(swept_meshes, ch=False, mergeUVSets=True, name="motion_structure")[0]
    pm.delete(full_torus, constructionHistory=True)
    pm.select(full_torus)
    pm.polyNormal(full_torus, normalMode=0, userNormalMode=0)

    light = create_light_torus(OUTLINE_RADIUS*RADIUS, 0.1, TORUS_SUBDIVISIONS, 25, "motion_structure_light")

    group = pm.group(em=True, name="motion_structure_group")
    pm.parent(full_torus, group)
    pm.parent(light, group)
    pm.xform(group, centerPivots=True)
    pm.delete(original_profile)

    return group



def create_outline(group, total=1):
    groupX = pm.group(em=True, name="structure_x_group")
    groupY = pm.group(em=True, name="structure_y_group")
    groupZ = pm.group(em=True, name="structure_z_group")
    for i in range(total):
          theta = i * (360 / total)
          
          meshX = pm.duplicate(group)[0]
          meshZ = pm.duplicate(group)[0]
          pm.parent(meshX, groupX)
          pm.parent(meshZ, groupZ)

          meshX.rename("motion_structure_X")
          meshZ.rename("motion_structure_Z")

          meshX.rotateX.set(theta)
          meshZ.rotate.set(0, 90, theta)

          if i != 0:
            meshY = pm.duplicate(group)[0]
            meshY.rename("motion_structure_Y")
            meshY.rotate.set(90, theta, 0)
            pm.parent(meshY, groupY)

    pm.delete(group)
    group2 = pm.group(em=True, name="motion_structure")
    pm.parent(groupX, group2)
    pm.parent(groupY, group2)
    pm.parent(groupZ, group2)
    return group2

def create_hypatia():
    radius = float(RADIUS * HYPATIA_RADIUS)
    sphere = pm.polySphere(name="hypatia", radius=radius, subdivisionsX=100, subdivisionsY=50)
    return sphere


def create_orbit():
    total = 7
    thickness = RADIUS * ORBIT_THICKNESS
    extrusion_thickness = RADIUS * ORBIT_THICKNESS
    height = RADIUS * ORBIT_HEIGHT
    min_radius = RADIUS / 5
    max_radius = RADIUS * (6/7)
    SM_AXIS_RATIO = 1 / math.sqrt(3)  # ~0.577
    group = pm.group(em=True, name="orbit_group")

    orbit_meshes = []
    ring_speeds = []
    planets = []

    for i in range(total):
        p = i / (total - 1) if total > 1 else 0
        ring_radius = min_radius + p * (max_radius - min_radius)
        perspective_factor_y = SM_AXIS_RATIO + p * (SM_AXIS_RATIO * (1 + SM_AXIS_RATIO) - SM_AXIS_RATIO)

        # Create a polyPlane as the base for our square profile.
        profile_poly = pm.polyPlane(width=height, height=height, subdivisionsX=1, subdivisionsY=1,
                                    name=f'ProfilePoly_{i}')[0]
        # Scale the plane to get the desired extrusion thickness.
        pm.xform(profile_poly, scale=(extrusion_thickness, 1, 1))
        # Rotate it so the face is oriented correctly for the sweep.
        pm.xform(profile_poly, rotation=(90, 90, 0))

        # Convert the polyPlane to a NURBS curve ("nurb square") by converting its border edges.
        edges = pm.polyListComponentConversion(profile_poly, toEdge=True)
        pm.select(edges, r=True)
        profile_curve = pm.polyToCurve(form=2, degree=1, name=f'ProfileCurve_{i}')[0]
        pm.delete(profile_poly)
        pm.select(clear=True)

        circle_path = pm.circle(radius=ring_radius, normal=(0, 1, 0), 
                                sections=TORUS_SUBDIVISIONS, name="torusPath")[0]
        pm.xform(circle_path, scale=(perspective_factor_y, 1, 1))  # Apply elliptical scaling
        
        # Align the profile: move it so its origin sits at the torus edge.
        pm.xform(profile_curve, translation=(ring_radius, 0, 0))
        # Rotate the profile so that it's perpendicular to the path.
        pm.xform(profile_curve, rotation=(0, 90, 0), worldSpace=True)
        
        # Perform the sweep (extrude in pipe mode).
        swept = pm.extrude(profile_curve, circle_path, et=2, upn=True, 
                        fixedPath=True, scale=1, useComponentPivot=True, 
                        name="sweptTorus")[0]
        
        poly_mesh = pm.nurbsToPoly(swept, format=3, polygonType=1, name="polySweptTorus")[0]
        pm.polySoftEdge(poly_mesh, angle=0, ch=1)

        pm.delete(swept)
    
        # Cleanup the construction curves.
        pm.delete(profile_curve, circle_path)

        # Center the resulting swept mesh.
        bbox = pm.exactWorldBoundingBox(poly_mesh)
        center = ((bbox[0] + bbox[3]) / 2.0,
                  (bbox[1] + bbox[4]) / 2.0,
                  (bbox[2] + bbox[5]) / 2.0)
        pm.move(-center[0], -center[1], -center[2], poly_mesh, r=True)
        pm.xform(poly_mesh, centerPivots=True)
        # pm.rotate(poly_mesh, lerp(45, 270, p), 0, lerp(0, 270, p))

            # --- Separate the inner side of poly_mesh ---
        # Here we calculate the full 3D distance from (0,0,0) for each face's center.

        mm = pm.PyNode(poly_mesh)

        # Duplicate the mesh
        inner_side = pm.duplicate(mm, name="inner")[0]
        outer_side = pm.duplicate(mm, name="outer")[0]

        # Define face range (adjust dynamically if needed)
        face_count = len(mm.faces)-1

        # Delete inner faces from the outer mesh
        pm.delete(outer_side.f[0:599])

        # Delete outer faces from the inner mesh
        pm.delete(inner_side.f[600:face_count])

        ringGroup = pm.group(em=True, name="orbit")

        # Parent both parts to the group
        pm.parent(inner_side, ringGroup)
        pm.parent(outer_side, ringGroup)

        pm.parent(ringGroup, group)

        # Cleanup
        pm.delete(mm)  
                
        # orbit_meshes.append(poly_mesh)
        # pm.parent(poly_mesh, group)

        # UV Mapping using Cylindrical Projection
        adjust_uvs_to_range(outer_side, projection_scale_u=1, projection_scale_v=1)
        adjust_uvs_to_range(inner_side, projection_scale_u=1, projection_scale_v=1)

    pm.select(clear=True)
    return orbit_meshes, ring_speeds, planets



# distribute_cubes_on_cube(big_size=500, grid=10, small_dims=(1, 1, 50.0))
# create_outline(create_motion_orbit_sculpture(), total=5)
create_orbit()
# create_hypatia()